src/EntityComponentManager.cc

```diff
@@ -270,6 +276,154 @@ Entity EntityComponentManagerPrivate::CreateEntityImplementation(Entity _entity)
+
+  for (const auto &[clonedModel, oldCanonicalLink] : oldModelCanonicalLink)
+  {
+    auto iter = oldToClonedCanonicalLink.find(oldCanonicalLink);
+    if (iter == oldToClonedCanonicalLink.end())
+    {
+      ignerr << "Error: attempted to clone model(s) with canonical link(s), "
+        << "but entity [" << oldCanonicalLink << "] was not cloned as a "
+        << "canonical link." << std::endl;
+      continue;
+    }
+    const auto clonedCanonicalLink = iter->second;
+    this->SetComponentData<components::ModelCanonicalLink>(clonedModel,
+        clonedCanonicalLink);
+  }
+
+  return clonedEntity;
+}
+
+Entity EntityComponentManager::Clone(const Entity _entity, const Entity _parent,
+    const std::string &_name, bool _allowRename,
+    std::unordered_map<Entity, Entity> &_oldModelCanonicalLink,
+    std::unordered_map<Entity, Entity> &_oldToClonedCanonicalLink)
+{
+  auto uniqueNameGenerated = false;
+
+  // Before cloning, we should make sure that:
+  //  1. The entity to be cloned exists
+  //  2. We can generate a unique name for the cloned entity
+  if (!this->HasEntity(_entity))
+  {
+    ignerr << "Requested to clone entity [" << _entity
+      << "], but this entity does not exist." << std::endl;
+    return kNullEntity;
+  }
+  else if (!_name.empty() && !_allowRename)
+  {
+    if (kNullEntity != this->EntityByComponents(components::Name(_name)))
+    {
+      ignerr << "Requested to clone entity [" << _entity
+        << "] with a name of [" << _name << "], but another entity already "
+        << "has this name." << std::endl;
+      return kNullEntity;
+    }
+    uniqueNameGenerated = true;
+  }
+
+  auto clonedEntity = this->CreateEntity();
+
+  if (_parent != kNullEntity)
+  {
+    this->SetParentEntity(clonedEntity, _parent);
+    this->CreateComponent(clonedEntity, components::ParentEntity(_parent));
+  }
+
+  // make sure that the cloned entity has a unique name
+  auto clonedName = _name;
+  if (!uniqueNameGenerated)
+  {
+    if (clonedName.empty())
+    {
+      auto originalNameComp = this->Component<components::Name>(_entity);
+      clonedName =
+        originalNameComp ? originalNameComp->Data() : "cloned_entity";
+    }
+    uint64_t suffix = 1;
+    while (kNullEntity != this->EntityByComponents(
+          components::Name(clonedName + "_" + std::to_string(suffix))))
+      suffix++;
+    clonedName += "_" + std::to_string(suffix);
+  }
+  this->CreateComponent(clonedEntity, components::Name(clonedName));
+
+  // copy all components from _entity to clonedEntity
+  for (const auto &type : this->ComponentTypes(_entity))
+  {
+    // skip the Name and ParentEntity components since those were already
+    // handled above
+    if ((type == components::Name::typeId) ||
+        (type == components::ParentEntity::typeId))
+      continue;
+
+    auto originalComp = this->ComponentImplementation(_entity, type);
+    auto clonedComp = originalComp->Clone();
+
+    if (type == components::Light::typeId)
```

Comment:
Handling each component type's special case within the ECM doesn't scale well. How about overriding the `Component::Clone()` function that you created so each component can handle its cloning special needs separately?

In any case, I'd say we don't need to worry about updating the contents of SDF components when cloning. These contents become outdated the moment the `SdfEntityCreator` generates components for them and it's not feasible to keep everything in sync. We can treat most SDF components as the "initial SDF" for an entity.

Reply:
> How about overriding the Component::Clone() function that you created so each component can handle its cloning special needs separately?

That sounds like a good approach :+1: 

> In any case, I'd say we don't need to worry about updating the contents of SDF components when cloning.

The issue is that if I don't update the name of the SDF for lights and actors, then they won't get created because another light/actor already exists with this name - in fact, the `SceneManager` will crash if the name isn't changed. Take a look at this code as an example:
https://github.com/ignitionrobotics/ign-gazebo/blob/1334b5cb6a123cd16c5d39b1cebb03e2f25a2bee/src/rendering/SceneManager.cc#L1111-L1117

If I try cloning a light without changing the name in the SDF, the `CreateSpotLight` call returns a `nullptr` because a light already exists with that name, and then that `nullptr` is dereferenced a few lines later, causing a crash. I can check to make sure that the name used for cloning something isn't already taken, but then I get into the issues described in https://github.com/ignitionrobotics/ign-gazebo/pull/959#issuecomment-919454527.

So, should I still modify the name in the SDF for lights and actors if I were to override `Component::Clone()` for these components?

Reply:
> The issue is that if I don't update the name of the SDF for lights and actors, then they won't get created because another light/actor already exists with this name

Argh I see :woman_facepalming: I think we shouldn't be using the `sdf::Light::Name` there, but the `Name` component.

So you can either override `Component::Clone` or update `SceneManager` to use the name component. Whatever is faster for now :+1:

Reply:
The `SceneManager` has been updated to use `components::Name`: https://github.com/ignitionrobotics/ign-gazebo/pull/959/commits/cef373f156ebd36abbf87d59aa5025fd37487bb2

I believe that resolves this conversation, unless there's anything else I am missing?